### ISSUE TITLE
[Google AI Studio] Gemini API without the API key

### ISSUE THREAD

**Issue by iG8R** [xtekky/gpt4free#1949]
Is there any way to implement getting access to `Google AI Studio` https://aistudio.google.com/app/prompts/new_chat, which is essentially the Gemini API but without an API key, similar to how OpenAI Chat only requires a `har` file but not an API key?

Why such a request appeared - recently, I don't know what happened with Google, but it banned all existing and newly created API keys and projects (since May 07, 2024 - 4 projects with 8 API keys) for a ridiculous reason, supposedly there was an exceeding of the quota metric 'Generate Content API requests per minute' and the limit 'GenerateContent request limit per minute for a region' of the service 'generativelanguage.googleapis.com" for a consumer project, although in the Quota&System Limits parameter of each project all `Current Usage Percentages` for all API services are equal to 0% and I know 100% that the application that used those API keys only made 1 request per 2 seconds.
To all my appeal requests I got nothing.

But `Google AI Studio` https://aistudio.google.com/app/prompts/new_chat still works fine.

![image](https://github.com/xtekky/gpt4free/assets/11407417/5096da0d-d33d-4b55-b28c-48ca00944f4f)

![image](https://github.com/xtekky/gpt4free/assets/11407417/2f951ed6-f997-4beb-8ca2-79f4bcb0b2cc)

**Comment by iG8R** [comment]
One more reason to use `Google AI Studio`

![image](https://github.com/xtekky/gpt4free/assets/11407417/da8f14da-7f1f-4d91-8f40-6feb750730e5)

**Comment by iG8R** [comment]
Up.

**Comment by JayJayBinks** [comment]
I experimented a bit with the Google API and 1.5 flash has far better quota.
I want to create a PR to add the gemini flash version, but am not allowed to create a new branch. @xtekky ?

**Comment by Zedai00** [comment]
@JayJayBinks fork it then add and do a pull request

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by Ran-Mewo** [comment]
did anyone do this yet?

**Comment by iG8R** [comment]
@hlohaus 
Could you please reconsider reopening this request?

**Comment by zakiel2012** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by hackerdy** [comment]
Any updates?

**Comment by hlohaus** [comment]
Hey @iG8R, I checked out this service.  It makes hashes from the request and puts them in the header.  They check the hashes to make sure they're good and you can't reuse them. I haven't found the hashing algorithm yet. Is there another package on GitHub that works?

**Comment by iG8R** [comment]
@hlohaus 
Thank you very much for bringing this issue to your attention.
Unfortunately, there was only one project related to the very `aistudio.google.com` site - https://github.com/SreejanPersonal/Gemini-1.5-Pro-Google-AI-Studio-Reverse-Engineered-API - but, now it is no longer working and abandoned, most likely precisely because of such difficulties in finding the hashing algorithm.
But as I've discovered, `aistudio.google.com`... how shall I put it... is more advanced than `gemini.google.com` in the sense that it can accommodate more context, and the models there, sort of, work more accurately, precisely, and clearly. At least, I see it in the translations of various texts that both sites perform simultaneously with the same texts.

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by Ran-Mewo** [comment]
I wonder when someone will reverse engineer it and get it working

**Comment by hlohaus** [comment]
I have developed a nodriver script for the video engine; however, the free usage limitations render reserving the Google AI engine impractical.

See the Video Provider.

**Comment by iG8R** [comment]
@hlohaus 
I don't know about the video usage in AI Studio, but I use it almost every day with mixed content (text, images, files - txt, pdf) and I already have about 50 chats with the amount of tokens ranging from about 720.000 to 1.000.000 (yeah, yeah, it is lagging terribly).
And I've never ever seen any usage limitations, there were about 10-15 times when it crashed with errors, something like "Unable count tokens" or "Unable save chat", but that is all.

**Comment by unixfox** [comment]
Alternative take: use the unofficial API in the new Gemini CLI

**Comment by iG8R** [comment]
@unixfox 
The unofficial API in the new Gemini CLI, as far as I know, uses gemini.google.com, and I'm already using it too, but it differs significantly compared to aistudio.google.com in terms of context size, time for new models to be added, and, quite strangely, inference quality.

**Comment by iG8R** [comment]
up

**Comment by iG8R** [comment]
up

**Comment by xer-on** [comment]
hello, i am new to this but i wanted to know that if i reverse engineer google or any ai provider like new kimi k2s frontend chat application's api can i use it in my apps with tool calling?

**Comment by OrionCodeDev** [comment]
More info i posted in #3092 . I think it will help to create a new provider in gpt4free

**Comment by iG8R** [comment]
@OrionCodeDev 
Thank you very much!

**Comment by OrionCodeDev** [comment]
@hlohaus it will help or you need more debug infos?

**Comment by hlohaus** [comment]
Hey, can you show me how to make the headers? I tried to decrypt some stuff. It use RC5 i thing. @OrionCodeDev

**Comment by OrionCodeDev** [comment]
> Hey, can you show me how to make the headers? I tried to decrypt some stuff. It use RC5 i thing. [@OrionCodeDev](https://github.com/OrionCodeDev)

@hlohaus 
Here is M1noa notes:
https://github.com/M1noa/how/blob/main/AIStudio-API.md
Here is my reverse try notes, where i be really close to replicate it, but that botguard token is too hard to decrypt/fight:
https://www.markdownpaste.com/document/370e55

**Comment by hlohaus** [comment]
Your LLM hallucinates. There is no botguard token i think.
I need this hash's in the Authorization header:

SAPISIDHASH 1753900196_{hash} SAPISID1PHASH 1753900196_{hash} SAPISID3PHASH 1753900196_{hash}

**Comment by hlohaus** [comment]
Hey, I added the Google Gemini CLI provider to the main branch! You can use gemini-pro for free in GPT4Free now. @iG8R

**Comment by iG8R** [comment]
@hlohaus 
Trying to use `Google Gemini CLI` I always get `An error occurred: RuntimeError: GCP_SERVICE_ACCOUNT environment variable not set.` Doen't it work via `nodriver`?

**Comment by hlohaus** [comment]
To begin, please install and authenticate with the gemini-cli tool. Subsequently, the system will utilize its authentication file. @iG8R

**Comment by KaiSoKei** [comment]
> Your LLM hallucinates. There is no botguard token i think. I need this hash's in the Authorization header:
> 
> SAPISIDHASH 1753900196_{hash} SAPISID1PHASH 1753900196_{hash} SAPISID3PHASH 1753900196_{hash}


yes i had faced that issue i just tried to build chrome exstention to talk to the ai as their ui is lagging as hell and they not willing to fix it 
i came accros the same issue and thats how i found this issue trying to get some help

**Comment by hlohaus** [comment]
@KaiSoKei Did you try to use the Gemini CLI provider with gpt4free?

**Comment by ezyyeah** [comment]
Hey guys I just made a relay for AI Studio, which "exposes" it as an API.

https://x.com/ezyyeah1/status/1961582725015408665

The chrome instances are workers (simulating send, intercepting the GenerateContent), and a http server is the manager.

It is not public yet, but making it open source is in my plans.

**Comment by OrionCodeDev** [comment]
> Hey guys I just made a relay for AI Studio, which "exposes" it as an API.
> 
> https://x.com/ezyyeah1/status/1961582725015408665
> 
> The chrome instances are workers (simulating send, intercepting the GenerateContent), and a http server is the manager.
> 
> It is not public yet, but making it open source is in my plans.

nice. waiting for open source.

**Comment by hlohaus** [comment]
Hey @OrionCodeDev,

You can use gemini-pro for free with the GeminiCLI provider. Just install and set up gemini-cli next to your gpt4free and use it with the "GeminiCLI" provider in gpt4free.
